jeff65/gold/ast.py

```diff
@@ -573,12 +618,33 @@ def describe(self):
         return f"(isr {repr(self.name)}\n    {stmts})"
 
 
-class PunctuationEndIsrNode(TokenNode):
+class PunctuationEndIsrNode(PunctuationEndNode):
+    pass
+
+
+class StatementWhileNode(TokenNode):
     def __init__(self, position, text):
-        super().__init__(Power.delimiter_endfun, position, text, end=True)
+        super().__init__(Power.statement, position, text)
+        self.condition = None
 
     def nud(self, right):
-        raise ParseError("unexpected 'endisr'", self)
+        self.condition = self.parse(right)
+        self.children = []
+        if type(right.current) is PunctuationDoNode:
+            right.next()
+            while type(right.current) is not PunctuationEndNode:
+                self.children.append(self.parse(right))
+            right.next()
+        else:
```

Comment:
this branch supports "shorthand" for single-statement loops without `do` and `end`, like:
```
while x < 5
    x += 1
```

i don't know if we actually want to allow that or not. if we do, it would also apply to `for`, `if`, `elseif`, and `else`.

i guess i have a slight preference _against_ allowing it because it saves a few keystrokes but is sometimes the source of refactoring mistakes. i put this together just to see what it would look like. i don't feel strongly one way or the other. what do you think, @jdpage?

Reply:
I have a fairly strong preference against allowing it. Multiple syntaxes for loops is (a) confusing, and (b) I am a member of the Braces Always school of thought.